src/main/java/com/mraof/minestuck/network/EditmodeDragPacket.java

```diff
@@ -0,0 +1,342 @@
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SoundType;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public final class EditmodeDragPacket
+{
+	private static final Logger LOGGER = LogManager.getLogger();
+	
+	private static boolean editModePlaceCheck(Level level, Player player, InteractionHand hand)
+	{
+		if(!level.isClientSide() && ServerEditHandler.getData(player) != null)
+		{
+			EditData data = ServerEditHandler.getData(player);
+			SburbConnection connection = data.getConnection();
+			
+			ItemStack stack = player.getMainHandItem();
+			
+			if(stack.isEmpty() || hand.equals(InteractionHand.OFF_HAND))
+				return false;
+			
+			DeployEntry entry = DeployList.getEntryForItem(stack, connection, level);
+			if(entry != null)
+			{
+				GristSet cost = entry.getCurrentCost(connection);
+				if(!GristHelper.canAfford(PlayerSavedData.getData(connection.getClientIdentifier(), level).getGristCache(), cost))
+					return false;
+				
+			}
+			else if(!(stack.getItem() instanceof BlockItem) || !GristHelper.canAfford(level, connection.getClientIdentifier(), GristCost.findCostForItem(stack, null, false, level)))
+				return false;
+		}
+		
+		return true;
+	}
+	
+	private static GristSet editModePlaceCost(Level level, Player player, InteractionHand hand, BlockPos pos)
+	{
+		EditData data = ServerEditHandler.getData(player);
+		SburbConnection connection = data.getConnection();
+		
+		ItemStack stack = player.getMainHandItem();
+		
+		if(stack.isEmpty() || hand.equals(InteractionHand.OFF_HAND))
+			return new GristSet();
+		
+		DeployEntry entry = DeployList.getEntryForItem(stack, connection, level);
+		if(entry != null)
+		{
+			GristSet cost = entry.getCurrentCost(connection);
+			if(level.getBlockState(pos).getMaterial().isReplaceable() && !GristHelper.canAfford(PlayerSavedData.getData(connection.getClientIdentifier(), level).getGristCache(), cost))
+				return cost;
+			
+		}
+		else if(level.getBlockState(pos).getMaterial().isReplaceable() && !GristHelper.canAfford(level, connection.getClientIdentifier(), GristCost.findCostForItem(stack, null, false, level)))
+			return GristCost.findCostForItem(stack, null, false, level);
+		
+		return new GristSet();
+	}
+	
+	private static boolean editModeDestroyCheck(Level level, Player player, BlockPos pos)
+	{
+		if(!level.isClientSide() && ServerEditHandler.getData(player) != null)
```

Comment:
These should already be checked by the caller of this function, so I recommend taking in the EditData as a parameter instead.